### PR TITLE
Add CDS template generation

### PR DESCRIPTION
**Pull request by Thiyanwso2**

### CODE REVIEW

File: native/cds-bal-template/src/main/resources/template/cdsPackageMd.vm [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+### Sample Config.toml file
+```
+[[ballerinax.health.fhir.cds.cds_services]]
+hook = "patient-view"
+title = "Static CDS Service Example"
+description = "An example of a CDS Service that returns a static set of cards"
+id = "static-patient-greeter"
+[ballerinax.health.fhir.cds.cds_services.prefetch]
+patientToGreet = "Patient/{{context.patientId}}"
+```
```
Comment (isuruh15) [inline native/cds-bal-template/src/main/resources/template/cdsPackageMd.vm:51]: newline is missing

File: native/health-cli/src/main/java/io/ballerina/health/cmd/HealthCmd.java [hunk 2]
```diff
@@ -19,17 +19,14 @@
...
 import io.ballerina.health.cmd.fhir.FhirSubCmd;
 import io.ballerina.health.cmd.hl7.Hl7SubCmd;
 import picocli.CommandLine;
 
-import java.io.BufferedReader;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
-import java.io.PrintStream;
+import java.io.*;
```
Comment (isuruh15) [inline native/health-cli/src/main/java/io/ballerina/health/cmd/HealthCmd.java:29]: remove wildcard. check in the other places as well.

File: native/health-cli/src/main/java/io/ballerina/health/cmd/core/config/HealthCmdConfig.java [hunk 3]
```diff
@@ -21,14 +21,15 @@
...
 import io.ballerina.health.cmd.core.exception.BallerinaHealthException;
+import net.consensys.cava.toml.Toml;
+import net.consensys.cava.toml.TomlParseResult;
 import org.apache.commons.lang.StringUtils;
 
-import java.io.FileNotFoundException;
-import java.io.FileReader;
-import java.io.InputStream;
-import java.io.InputStreamReader;
+import java.io.*;
```
Comment (isuruh15) [inline native/health-cli/src/main/java/io/ballerina/health/cmd/core/config/HealthCmdConfig.java:28]: remove this. look for other places as well.

File: native/health-cli/src/main/java/io/ballerina/health/cmd/core/utils/HealthCmdUtils.java [hunk 4]
```diff
@@ -96,8 +107,33 @@ public static InputStream getResourceFile(Class<?> handlerClass, String fileName
...
         }
         return ioStream;
     }
+
+    public static JsonObject parseTomlToJson(String path) {
+        TomlParseResult parseResult;
+        try {
+            parseResult = Toml.parse(Paths.get(path));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (isuruh15) [inline native/health-cli/src/main/java/io/ballerina/health/cmd/core/utils/HealthCmdUtils.java:120]: Shall we throw `BallerinaHealthException ` here?

File: native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2023, WSO2 LLC. (http://www.wso2.com).
```
Comment (isuruh15) [inline native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java:2]: update year

File: native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+                }
+                if (packageName != null && !packageName.isEmpty()) {
+                    JsonElement overrideConfig = new Gson().toJsonTree(packageName.toLowerCase());
+                    toolConfigInstance.overrideConfig(PROJECT_PACKAGE_NAME_PREFIX, overrideConfig);
+                }
+
+                Class<?> toolClazz = classLoader.loadClass(CDS_TOOL_CLASS_NAME);
+                tool = (Tool) toolClazz.getConstructor().newInstance();
+                tool.initialize(toolConfigInstance);
+                crdTemplateGenerator = tool.execute(null);
```
Comment (isuruh15) [inline native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java:136]: let's add a comment here mentioning the reason for passing null as the toolContext.

File: native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+                HealthCmdUtils.throwLauncherException(e);
+            } catch (InstantiationException | IllegalAccessException e) {
+                printStream.println(ErrorMessages.CONFIG_INITIALIZING_FAILED);
+                HealthCmdUtils.throwLauncherException(e);
+            } catch (CodeGenException e) {
+                printStream.println(ErrorMessages.UNKNOWN_ERROR);
+                printStream.println(e);
+                HealthCmdUtils.throwLauncherException(e);
+            } catch (InvocationTargetException | NoSuchMethodException e) {
+                throw new RuntimeException(e);
```
Comment (isuruh15) [inline native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java:148]: Can't we throw a Launcher Exception here? Check for other places as well.

File: native/health-cli/src/main/resources/ballerina-health-cds.help [hunk 8]
```diff
@@ -0,0 +1,100 @@
+NAME
```
Comment (isuruh15) [inline native/health-cli/src/main/resources/ballerina-health-cds.help:1]: this has only FHIR tools info. Have we missed something here?

File: native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/config/BallerinaProjectToolConfig.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+import org.wso2.healthcare.codegen.tool.framework.commons.Constants;
+import org.wso2.healthcare.codegen.tool.framework.commons.config.AbstractToolConfig;
+import org.wso2.healthcare.codegen.tool.framework.commons.exception.CodeGenException;
+import org.wso2.healthcare.codegen.tool.framework.commons.model.ConfigType;
+import org.wso2.healthcare.codegen.tool.framework.commons.model.JsonConfigType;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.wso2.healthcare.cds.codegen.ballerina.tool.CdsBallerinaProjectConstants.*;
```
[the remark this answers is not in the document]
Comment (Thiyanwso2) [reply native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/config/BallerinaProjectToolConfig.java]: Done

File: native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/config/DependencyConfig.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.healthcare.cds.codegen.ballerina.tool.config;
+
+import com.google.gson.JsonObject;
+
+import static org.wso2.healthcare.cds.codegen.ballerina.tool.CdsBallerinaProjectConstants.*;
```
[the remark this answers is not in the document]
Comment (Thiyanwso2) [reply native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/config/DependencyConfig.java]: Done

File: native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/config/BallerinaProjectToolConfig.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            }
+
+            jsonConfigObj = ((JsonConfigType) configObj).getConfigObj().getAsJsonObject(HOOKS);
+            loadCdsHooks(jsonConfigObj);
+        }
+    }
+
+    public void loadCdsHooks(JsonObject cdsHookConfig) {
+        if (cdsHookConfig.get(CDS_SERVICES) instanceof JsonObject) {
+            System.out.println(cdsHookConfig.getAsJsonObject(CDS_SERVICES));
```
[the remark this answers is not in the document]
Comment (Thiyanwso2) [reply native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/config/BallerinaProjectToolConfig.java:79]: Removed

File: native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/config/MetadataConfig.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+    private String org;
+    private String namePrefix;
+    private String version;
+    private String distribution;
+    private final List<String> authors = new ArrayList<>();
+    private final List<String> keywords = new ArrayList<>();
+
+    public MetadataConfig(JsonObject packageConfig) {
+        //todo: add null checks
```
[the remark this answers is not in the document]
Comment (Thiyanwso2) [reply native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/config/MetadataConfig.java:49]: Done

File: native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/generator/BallerinaProjectGenerator.java [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.wso2.healthcare.cds.codegen.ballerina.tool.CdsBallerinaProjectConstants.*;
+
+/**
+ * Generator class to wrap all the generator classes in Ballerina project generator.
+ */
+public class BallerinaProjectGenerator extends AbstractFHIRTemplateGenerator {
```
[the remark this answers is not in the document]
Comment (Thiyanwso2) [reply native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/generator/BallerinaProjectGenerator.java:53]: Renamed

File: native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/model/CdsHook.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.wso2.healthcare.cds.codegen.ballerina.tool.CdsBallerinaProjectConstants.*;
+
+public class CdsHook {
+    String id;
```
[the remark this answers is not in the document]
Comment (Thiyanwso2) [reply native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/model/CdsHook.java:36]: Done

File: native/health-cli/src/main/java/io/ballerina/health/cmd/HealthCmd.java (hunk 2 above)
Comment (Thiyanwso2) [reply native/health-cli/src/main/java/io/ballerina/health/cmd/HealthCmd.java]: Done

File: native/health-cli/src/main/java/io/ballerina/health/cmd/core/config/HealthCmdConfig.java (hunk 3 above)
Comment (Thiyanwso2) [reply native/health-cli/src/main/java/io/ballerina/health/cmd/core/config/HealthCmdConfig.java]: Done

File: native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java (hunk 5 above)
Comment (Thiyanwso2) [reply native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java:2]: Done

File: native/health-cli/src/main/resources/ballerina-health-cds.help (hunk 8 above)
Comment (Thiyanwso2) [reply native/health-cli/src/main/resources/ballerina-health-cds.help]: Updated

File: native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java (hunk 6 above)
Comment (Thiyanwso2) [reply native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java:147]: Done

File: native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/model/CdsHook.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (Thiyanwso2) [reply native/cds-bal-template/src/main/java/org/wso2/healthcare/cds/codegen/ballerina/tool/model/CdsHook.java]: Done

File: native/health-cli/src/main/java/io/ballerina/health/cmd/HealthCmd.java (hunk 2 above)
Comment (Thiyanwso2) [reply native/health-cli/src/main/java/io/ballerina/health/cmd/HealthCmd.java]: Done

File: native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java (hunk 7 above)
Comment (Thiyanwso2) [reply native/health-cli/src/main/java/io/ballerina/health/cmd/handler/CrdTemplateGenHandler.java:159]: Done

File: native/health-cli/src/main/java/io/ballerina/health/cmd/core/utils/HealthCmdUtils.java (hunk 4 above)
Comment (Thiyanwso2) [reply native/health-cli/src/main/java/io/ballerina/health/cmd/core/utils/HealthCmdUtils.java:120]: Added BLauncherException

**Review by isuruh15 — DISMISSED** [review]

File: native/cds-bal-template/src/main/resources/template/cdsInterceptor.vm [hunk 15]
```diff
@@ -0,0 +1,69 @@
+import ballerina/http;
```
Comment (sameeragunarathne) [inline native/cds-bal-template/src/main/resources/template/cdsInterceptor.vm:1]: Add the licence content as well to the templates

File: native/cds-bal-template/src/main/resources/template/cdsInterceptor.vm [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+import ballerina/lang.regexp;
+import ballerinax/health.fhir.cds;
+
```
Comment (sameeragunarathne) [inline native/cds-bal-template/src/main/resources/template/cdsInterceptor.vm:4]: Should we indicate that this file is autogenerated and not subject to be edited?

File: native/cds-bal-template/src/main/resources/template/cdsInterceptor.vm (hunk 15 above)
Comment (Thiyanwso2) [reply native/cds-bal-template/src/main/resources/template/cdsInterceptor.vm:19]: Added

File: native/cds-bal-template/src/main/resources/template/cdsInterceptor.vm (hunk 16 above)
Comment (Thiyanwso2) [reply native/cds-bal-template/src/main/resources/template/cdsInterceptor.vm:21]: Done

**Review by sameeragunarathne — APPROVED** [review]

**Pull request merged**